python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,601 @@
+
+
+class DoubleLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(DoubleLiteral, self).__init__(f"DoubleLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(DoubleType)
+    def _(self, type_var):
+        return self
+
+    @to.register(FloatType)
+    def _(self, type_var):
+        if FloatType.max < self.value:
+            return AboveMax()
+        elif FloatType.min > self.value:
+            return BelowMin()
+        return FloatLiteral(self.value)
+
+    @to.register(DecimalType)
+    def _(self, type_var):
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal("1."), rounding=ROUND_HALF_UP))
+        else:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP))
+
+
+class DateLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(DateLiteral, self).__init__(f"DateLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(DateType)
+    def _(self, type_var):
+        return self
+
+
+class TimeLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(TimeLiteral, self).__init__(f"TimeLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(TimeType)
+    def _(self, type_var):
+        return self
+
+
+class TimestampLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(TimestampLiteral, self).__init__(f"TimestampLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(TimestampType)
+    def _(self, type_var):
+        return self
+
+    @to.register(DateType)
+    def _(self, type_var):
+        return DateLiteral((datetime.datetime.fromtimestamp(self.value / 1000000) - EPOCH).days)
+
+
+class DecimalLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(DecimalLiteral, self).__init__(f"DecimalLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(DecimalType)
+    def _(self, type_var):
+        if type_var.scale == abs(self.value.as_tuple().exponent):
+            return self
+        return None
+
+
+class StringLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(StringLiteral, self).__init__(f"StringLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(StringType)
+    def _(self, type_var):
+        return self
+
+    @to.register(DateType)
+    def _(self, type_var):
+        import dateutil.parser
+
+        return DateLiteral((dateutil.parser.parse(self.value) - EPOCH).days)
```

Comment:
These conversions need to be very strict. A date must be formatted `YYYY-mm-DD`, so I think this should explicitly use a format and fail if the string deviates from the format in anyway.

Same thing for timestamp and timestamptz conversions. They should be ISO-8601 (with the `T`). Timestamp must not have any zone offset or additional information, and timestamptz must have a valid zone offset and not a zone name or descriptor.

Reply:
You may want to omit these for now and do them in a follow-up. I think it would make sense to have a set of util methods for date/time like the Java implementation has that are separately tested and used in multiple places.